### ISSUE TITLE
typescript defs missing `id` attribute for Box

### ISSUE THREAD

**Issue by toddself** [primer/components#540]
Trying to do:

```ts
      <Box
        key={marker.points}
        display="inline-block"
        id={marker.label}
        width={`${width}%`}
        bg={filled ? '#959da5' : ''}
      >
        <div style={dotStyle} />
      </Box
```

And it is complaining about `id`: 
```
Type '{ children: Element; key: number; display: string; id: string; width: string; bg: string; }' is not assignable to type 'IntrinsicAttributes & BoxProps & { children?: ReactNode; }'.
  Property 'id' does not exist on type 'IntrinsicAttributes & BoxProps & { children?: ReactNode; }'.
```

**Comment by dmarcey** [comment]
I think we have a couple of different options for a fix here.

1. Add `id` to `BoxProps` - seems not the best, eventually someone will need an id on a Button, etc.
2. Add `id` to `BaseProps` - better, we don't have to add id everywhere, but the next prop we want to add (e.g. `tabIndex` or `onClick`), we have to go back and add again.
3. Have `BoxProps` extend `React.HTMLAttributes<HTMLDivElement>` - since we're passing all of these props along to the div anyway, we might as well expose them on the props interface. We'd probably want to do the same for other interfaces. Today you can't pass `target='_blank'` to a `<Link />` or `spellcheck='true'` to a `<TextInput />`. My only concern with this approach is if there are any props which explicitly don't want to expose to pass along to the underlying element (`styles` comes to mind?). If that's the case, then we can figure out a way to omit those from the type definition.

@emplums / @colebemis  any thoughts?

**Comment by dmarcey** [comment]
Thanks, @T-Hugs . The more interesting challenge might be figuring out a good way to support the `as` prop.

`as` is used to use a different underlying component as the "base" component for what you're trying to render (https://primer.style/components/docs/core-concepts). 

`<Button as="a" href="https://github.com" />`

Right now, I don't think this will work for typescript, because the type of the `as` prop is just 
`as?: React.ReactType` and href isn't defined on ButtonProps.

It would be interesting to see if there was some way(using generics, I'm guessing) to be able to get some type safety based on the type supplied in the `as` prop.

**Comment by dmarcey** [comment]
Fixed in `14.0.0` #501

**Comment by toddself** [comment]
Thanks y'all!
